app/src/main/java/com/UserDatabase/UserDatabase.java

```diff
@@ -0,0 +1,264 @@
+package com.UserDatabase;
+
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.reflect.TypeToken;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+
+import static org.junit.Assert.assertEquals;
+
+public abstract class UserDatabase {
+    protected final String LogTag = "Database";
+
+    // Path of database file
+    private final File Filepath;
+
+    // Database identifier, to ensure safe loading
+    private final String Id;
+    // Vector target length, to validate database inputs
+    private final int VectorLength;
+
+    // protected ArrayList<UserRecord> usersRecords;
+    protected Map<String, float[]> usersRecords;
+
+
+    UserDatabase(File filepath, String id, int vectorLength) {
+        this.Filepath = filepath;
+        this.Id = id;
+        this.VectorLength = vectorLength;
+
+        usersRecords = new HashMap<String, float[]>();
+
+        // Load Database on creation
+        loadDatabase();
+    }
+
+    /**
+     * Find closest record from the database. Algorithm and time complexity is dependent
+     * on database type.

```

Comment:
idk how complexity of the algorithm may be depended on database in this case, if you think about having param vector mapped and sorted, it might be better to describe it on certain implementation. In my opinion it has nothing to do with database in this case.

Reply:
+1